drivers/serial/uart_stm32.c

```diff
@@ -2014,6 +2014,105 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
 #define STM32_UART_PM_WAKEUP(index) /* Not used */
 #endif
 
+/* Ensure DTS doesn't present an incompatible parity configuration.
+ * Mark/space parity isn't supported on the STM32 family.
+ * If 9 data bits are configured, ensure that a parity bit isn't set.
+ */
+#define STM32_UART_CHECK_DT_PARITY(index)				\
+BUILD_ASSERT(								\
+	!(DT_INST_ENUM_IDX_OR(index, parity, STM32_UART_DEFAULT_PARITY)	\
+					== UART_CFG_PARITY_MARK ||	\
+	DT_INST_ENUM_IDX_OR(index, parity, STM32_UART_DEFAULT_PARITY)	\
+					== UART_CFG_PARITY_SPACE),	\
+	"Node " DT_NODE_PATH(DT_DRV_INST(index))			\
+	" has unsupported parity configuration");			\
+BUILD_ASSERT(								\
+	!(DT_INST_ENUM_IDX_OR(index, parity, STM32_UART_DEFAULT_PARITY)	\
+					!= UART_CFG_PARITY_NONE	&&	\
+	DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			    STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_9),	\
+	"Node " DT_NODE_PATH(DT_DRV_INST(index))			\
+		" has unsupported parity + data bits combination");
+
+/* Ensure DTS doesn't present an incompatible data bits configuration
+ * The STM32 family doesn't support 5 or 6 data bits, only some series support 7
+ */
+#ifdef LL_USART_DATAWIDTH_7B
+#define STM32_UART_CHECK_DT_DATA_BITS(index)				\
+BUILD_ASSERT(								\
+	!(DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			      STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_5 ||	\
+	DT_INST_ENUM_IDX_OR(index, data_bits,				\
+			    STM32_UART_DEFAULT_DATA_BITS)		\
+					== UART_CFG_DATA_BITS_6),	\
```

Comment:
6-bit payload with a parity bit is possible on STM32 (that's a valid case of LL_USART_DATAWIDTH_7B), so this condition should be added to the check.
Though the function `uart_stm32_cfg2ll_databits` is incomplete because it does not support the case 6-bit data + parity.

Reply:
Ahh so parity is included in the `LL_USART_DATAWIDTH` calculation? Looking at the HAL back when I created this PR, this wasn't clear:
```c
#define LL_USART_DATAWIDTH_7B      ...    /*!< 7 bits word length : Start bit, 7 data bits, n stop bits */

```

I'll remove the exclusion of 6-bit data, thanks! Though ideally we should probably check if we have a parity bit defined in the case of 6-bit data.

Reply:
Adapted the check to ensure that parity is set if 6-data bits are defined on `LL_USART_DATAWIDTH_7B`.